### PR TITLE
JDK-8313251: Add NativeLibraryLoad event

### PR DESCRIPTION
**Pull request by MBaesken**
Add a NativeLibraryLoad event that provides us more detail about shared lib/dll loads. This gives a time stamp and success + error details of the load operation. It enhances the already existing information we get from the existing NativeLibrary event (that periodically samples the native modules of the jvm process).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313251](https://bugs.openjdk.org/browse/JDK-8313251): Add NativeLibraryLoad event (**Enhancement** - P3)

### Reviewers
 * [Johannes Bechberger](https://openjdk.org/census#jbechberger) (@parttimenerd - Committer) ⚠️ Review applies to [da6b9861](https://git.openjdk.org/jdk/pull/15065/files/da6b9861a09ba577a90ae012b1321da031c703c7)
 * [Erik Gahlin](https://openjdk.org/census#egahlin) (@egahlin - **Reviewer**) ⚠️ Review applies to [045f2832](https://git.openjdk.org/jdk/pull/15065/files/045f2832536564aa6f5940fa737a47eab8e62839)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15065/head:pull/15065` \
`$ git checkout pull/15065`

Update a local copy of the PR: \
`$ git checkout pull/15065` \
`$ git pull https://git.openjdk.org/jdk.git pull/15065/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15065`

View PR using the GUI difftool: \
`$ git pr show -t 15065`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15065.diff">https://git.openjdk.org/jdk/pull/15065.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15065#issuecomment-1655250843)

### CODE REVIEW

File: test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java
Comment (dholmes-ora) [inline test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java:2]: New test should have single copyright year, even if copied from existing test. Thanks.

**Comment by MBaesken** [comment]
Hi Johannes, thanks for the review !

May I get a second review please?

**Comment by egahlin** [comment]
> Hi Johannes, thanks for the review !
> 
> May I get a second review please?

I can review, but I like replies to my comments first.

**Comment by MBaesken** [comment]
> > Hi Johannes, thanks for the review !
> > May I get a second review please?
> 
> I can review, but I like replies to my comments first.

Hi Erik , maybe I miss something?  What comment(s)  are you referring to ?

**Comment by egahlin** [comment]
> > > Hi Johannes, thanks for the review !
> > > May I get a second review please?
> > 
> > 
> > I can review, but I like replies to my comments first.
> 
> Hi Erik , maybe I miss something? What comment(s) are you referring to ?

Hi Matthias. Sorry, I forgot to press submit.

File: src/hotspot/os/bsd/os_bsd.cpp
Comment (MBaesken) [reply src/hotspot/os/bsd/os_bsd.cpp:1027]: Hi Erik , I removed the JFR with configure and attempted this (however on Linux) but for some reason I got build errors. So I better added the macros. (not sure why this happened, with some other JFR event living without INCLUDE_JFR seemed to  work as expected)

File: src/hotspot/share/jfr/metadata/metadata.xml
Comment (MBaesken) [reply src/hotspot/share/jfr/metadata/metadata.xml:942]: Hi Erik, I will adjust. I will use errorMessage because on some platforms we have already coding to describe shared lib load error situations in more detail (see for example os_linux.cpp).

File: test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java
Comment (MBaesken) [reply test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java:51]: Hi Erik, this was taken from test/jdk/jdk/jfr/event/runtime/TestNativeLibrariesEvent.java , I do not see in main there a try with resources.  What do you want to put in the try block exactly ? The new Recording()  ?

File: src/hotspot/os/bsd/os_bsd.cpp
Comment (egahlin) [reply src/hotspot/os/bsd/os_bsd.cpp:1027]: OK

File: test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java
Comment (egahlin) [reply test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java:51]: Yes.
try (Recording r = new Recording()) {
 ...
}
It's not used everywhere, but it seems proper to release chunk resources when the test is finished.

(I should probably go over all tests and change to avoid the pattern being copied.)

Comment (MBaesken) [reply test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java]: HI Erik, I adjusted the test as recommended .

File: src/hotspot/os/bsd/os_bsd.cpp
Comment (dholmes-ora) [reply src/hotspot/os/bsd/os_bsd.cpp:1027]: Just for completeness the various os_xxx.cpp files all use INCLUDE_JFR around event code.

File: test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java
Comment (dholmes-ora) [inline test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java:77]: We have `Platform.sharedLibraryExt()` to simplify this a bit. Unfortunately we don't have a `sharedLibraryPrefix()` to reduce it to one line. :(

File: test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java
Comment (dholmes-ora) [inline test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java:54]: I don't think loading libawt is a good choice - maybe libnet or libnio? Headless builds may not have libawt and I'd be concerned about loading it directly and running the OnLoad hooks.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Overall looks okay but I have a request for the test. Thanks.

File: test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java
Comment (dholmes-ora) [reply test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java:77]: Okay

File: test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java
Comment (MBaesken) [reply test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java:54]: libnet does not work, it gives us the error  'Native Library /testdir/images/jdk/lib/libnet.so already loaded in another classloader' . So we have to load something that is not there already.  But I agree, awt might not be the best choice.  What about libinstrument ?

Comment (dholmes-ora) [reply test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java:54]: Yeah that might work. If worse comes to worst, you could create a trivial library just or the test.

File: test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java
Comment (MBaesken) [reply test/jdk/jdk/jfr/event/runtime/TestNativeLibraryLoadEvent.java:77]: I created  https://bugs.openjdk.org/browse/JDK-8313670  .

**Comment by MBaesken** [comment]
David are you fine with the latest version? 
I checked the GHA build errors and they are not related.

**Review by dholmes-ora — APPROVED** [review]
Looks okay. Thanks

**Comment by MBaesken** [comment]
Hi David , thanks for the review !

/integrate

**Pull request closed without merging**
